### PR TITLE
Ensure tolerance is available to EqualityAdapters and is not silently ignored

### PR DESCRIPTION
**Pull request by mr-russ**
Tolerance has been silently ignored in cases where certain types of equality were requested.  It was also impossible for someone to implement an External Comparer that considered tolerance.  This enables that and automates most comparisons in a linear range, without the need for custom NUnit expansions.

Tolerance gained the concept of HasVariance to better determine if strict equals could be applied or not.  Within() calls that were previously ignored can now determine if it's possible to apply the tolerance, and if not an error is raised.

### CODE REVIEW

**Comment by mr-russ** [comment]
This related to #2870, #3361 and #4310.  Each mention issues about tolerance and equality.  I'm not convinced it fixes all of them, but it's relevant context for them.

**Comment by mr-russ** [comment]
@dotnet-policy-service agree company="photoSentinel"

**Comment by mr-russ** [comment]
I have removed the dynamic calculations of tolerance to remove the CSharp dependency on .NET Framework.
I will see if this is acceptable before learning how the project would allow the addition of the required dependency.

File: src/NUnitFramework/framework/Constraints/Tolerance.cs [hunk 1]
```diff
@@ -169,10 +169,14 @@ public Tolerance Ticks
...
         /// <summary>
-        /// Returns true if the current tolerance has not been set or is using the .
+        /// Returns true if the current tolerance has not been set or is using the default.
         /// </summary>
         public bool IsUnsetOrDefault => Mode == ToleranceMode.Unset;
 
+        /// <summary>
+        /// Returns true if the current tolerance varies from exact and default. Indicating tolerance needs processing.
+        /// </summary>
+        public bool HasVariance => Mode != ToleranceMode.Unset && this != Exact && this != Default;
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Constraints/Tolerance.cs:179]: Default should not be necessary as that has mode `Unset`.
Also, the match on `Exact` is a reference comparison, and won't match `Within(0)`.
But maybe we don't want to, if they want a perfect match, don't specify `Within` at all.

Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:179]: What is the reason for  `HasVariance` being slightly different then `!IsUnsertOrDefault`?
What it your use case?

**Review by mr-russ — COMMENTED** [review]
Thanks for the review, I'm keen to get clarity on the questions at the bottom before I submit fixes.  If it's just "an -> a", great.  I feel there is more to clarify on the Exact comparisons.

**Comment by mr-russ** [comment]
@manfred-brands thanks for your comments and looking into the Exact tolerance further.  I was typing too late last night and did not look it up properly.  Going from memory is not usually a good thing.

I've committed the grammar fix.  I couldn't discern any actions from the other series of comments we've had.  So if there are further items to change please let me know.

File: src/NUnitFramework/framework/Constraints/Comparers/EquatablesComparer.cs [hunk 2]
```diff
@@ -15,6 +15,8 @@ internal static class EquatablesComparer
         {
             if (equalityComparer.CompareAsCollection && state.TopLevelComparison)
                 return null;
+            if (tolerance != null && tolerance.HasVariance)
```
Comment (stevenaw) [inline src/NUnitFramework/framework/Constraints/Comparers/EquatablesComparer.cs:18]: It looks like this is is because equatablescomparer checks don't use tolerance and can't because the interface doesn't support it. While this makes sense, I'm also curious if there's a chance we think this could break someone's test as it causes the EquatablesComparer to be skipped in this case. I've tried quite a bit to find edge cases here using built-in types and it seems fairly sound. 

One area of concern here for me though is for types that use `IEquatable` for equality comparisons between two different types, such as `string` and `Microsoft.Extensions.Primitives.StringSegment`. Such a test would require using the IEquatable comparer to pass, but the test would also currently pass if a tolerance were given, for example by appending `.Within(0)` .

```
[Test]
public void TestEquatables()
{
    var a = new Microsoft.Extensions.Primitives.StringSegment("a");
    var b = "a";
    Assert.That(a, Is.EqualTo(b).Within(0));
}
```

This change would cause such a test to fail with the message `"Tolerance is not supported for this comparison"`. I feel like this could be a good thing as it alerts about unnecessary comparison modifiers, but someone else may argue "if it doesn't apply why can't it be ignored?"

I admit the risk of this is low, but I'd be curious to hear other thoughts from the @nunit/framework-team about it. I think the risk here would be higher for custom types than for BCL types. In fact, this was the _only_ case I found in the BCL as most others would still pass using one of the other comparers in the comparer chain.

**Review by stevenaw — CHANGES_REQUESTED** [review]
Thanks @mr-russ it's clear that a lot of thought and effort went into this PR and the approach. I left a few comments, mostly around trying to avoid breaking changes with the hope that something like this may find it's way into a 3.13 release. Most would be easy to adjust, but I'd be curious to hear some thoughts from other team members on the tolerance check in the EquatablesComparer before you go to any trouble making changes there.

**Comment by stevenaw** [comment]
@jnm2 do you mind if I ask you for your thoughts on this change? I find you often have good insights into the chance of accidentally breaking changes and I suspect @mr-russ may hope this change finds it's way into a 3.13 release.

File: src/NUnitFramework/framework/Constraints/Comparers/EquatablesComparer.cs (hunk 2 above)
Comment (stevenaw) [reply src/NUnitFramework/framework/Constraints/Comparers/EquatablesComparer.cs:18]: Having slept on this, I think I understand the fix here and potential for discussion to be if we want a tolerance applied in an invalid situation to alert the user and fail the test or continue being ignored. I don't have much concern if it goes into v4, though perhaps it could be surprising or disruptive in 3.13 for someone to upgrade and find a test that once passed were to start failing. In either case, I'm curious the thoughts of the team

File: src/NUnitFramework/framework/Constraints/Tolerance.cs [hunk 3]
```diff
@@ -21,14 +21,14 @@ public sealed class Tolerance
         private const string NumericToleranceRequired = "A numeric tolerance is required";
 
         /// <summary>
-        /// Returns a default Tolerance object, equivalent to an exact match.
+        /// Returns a default Tolerance object, equivalent to a default matching rules.
         /// </summary>
-        public static Tolerance Default => new Tolerance(0, ToleranceMode.Unset);
+        public static readonly Tolerance Default = new Tolerance(0, ToleranceMode.Unset);
```
[the remark this answers is not in the document]
Comment (mr-russ) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:26]: Not as easily as I would like.  I'll need suggestions on how to equate Two tolerances, as they don't have an Equals() operator now.  Are you happy to have Equals implemented for Tolerance?  I'm not 100% sure how I create equal scenario, I'm assuming I'm testing both the type of the value, and that it thinks it's "Equal" to the value of the Exact tolerance.

I converted to this so that I can reference Equals and be sure Default, or Unset where what's really requested.  Maybe the other way where any submission of 0, Linear is considered exact is correct.

What are your thoughts on direction?

File: src/NUnitFramework/framework/Constraints/Tolerance.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mr-russ) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:179]: It's public because my default was to use what's used for properties of the same usage/type.  So I copied IsUnsetOrDefault.  Also other Tolerance implementers might want to be able to find out if the Tolerance HasVariance or not.  Especially useful for comparers.  If it's internal, how would a custom AreEqual constraint check if there is a variance?

In my many different attempts to write NodaTime extensions for NUnit, I have had to replace the tolerance with a custom implementation, as well as the EqualConstraint to make it work.  Extending Equals has been much more difficult that I would have hoped.  It's difficult to add an automatic addition for a new type.  Restricting access to this value makes maintenance easier, but I believe limits those who wish to extend NUnit.

I'm not sure how much extending of NUnit goes on, but my first attempt ran into a number of things that made it harder than I thought it would be.

File: src/NUnitFramework/framework/Constraints/Tolerance.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (OsirisTerje) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:26]: There are two objectives here,  "Ensure tolerance is available to EqualityAdapters" and "Is not silently ignored".   The first one is very fine.  For the second we do have some options:  What we have done earlier is 1) Things that are "wrong" or "not best practice", are handled by the NUnit.Analyzers and warned there with suitable refactoring added.   2) In the adapter project I have added feature flags around every possibly breaking change, making it default off in the upcoming version, and when it seems like people have got it, I have switched to default on. Sending flags down to the framework has not been so common though, but could be used. "ToleranceMode:  Strict or Relaxed".   A third option would be not to throw an exception but instead, output a warning/error message.    I think the latter might be best suited here, and in addition, ask @mikkelbu about adding a check to the analyzers.

[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:26]: Changing back to a property with just reference compare can be done like:
```
public static Tolerance Default { get; } = new Tolerance(0, ToleranceMode.Unset);
```

[the remark this answers is not in the document]
Comment (manfred-brands) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:26]: @OsirisTerje an nunit.analyzer to warn that tolerance will not be used with a codefix to remove the `.Within(...)` is a good idea. The issue already [exists](https://github.com/nunit/nunit.analyzers/issues/344) but it has stalled. I will pick it up myself.

[the remark this answers is not in the document]
Comment (mr-russ) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:26]: That's a very good point @OsirisTerje, and my discussion with @stevenaw about backpatching probably centres around being able to do (1), EqualityAdapters with tolerance support.  That's what's needed for more flexible extension.

My correctness bent is the surprise, it's been like this for a long while without complaint. So only making the tolerance checking changes in master (v4) is a good direction.

**Comment by OsirisTerje** [comment]
@stevenaw It seems this can be merged in now, what do you think?   Been a week now, so I don't think @jnm2 will have time this time.

**Comment by stevenaw** [comment]
Thanks for following up @OsirisTerje 
Sorry @mr-russ work got pretty busy for me this last week. I'm not at a computer at the moment, but I aim to get back to this within next 24hr.

File: src/NUnitFramework/framework/Constraints/Tolerance.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevenaw) [reply src/NUnitFramework/framework/Constraints/Tolerance.cs:179]: Thanks, that's a good point. I hadn't thought about custom comparers but they're certainly relevant even within the scope of this issue. That makes sense to me, thanks!

As a general rule, I'd suggest "internal by default" because public members are much more difficult to change in a non-breaking way, but I can see why this should be public. 👍

**Review by stevenaw — APPROVED** [review]
Thanks for everyone's patience for me to get back to this. Busy week for me as I've been wrapping up a few things.

Sounds like these changes are good to go into v4, but we may desire a more scaled-back version of the changes for 3.x, one which will pass tolerance to EqualityAdapters but will continue to silently ignore in cases where it's not needed. 

Thanks for your contribution @mr-russ ! Would you be willing to also put up a PR for the v3.x changes?

**Comment by OsirisTerje** [comment]
@mr-russ In your opinion, which of the issues #2870, #3361 and #4310 would you say is fixed by this PR?

**Comment by mr-russ** [comment]
I believe #2870 should be closed.
#3316 (I posted the wrong number via typo previously) requires, what I would like as the extension or using dynamic typing and runtime failures if +/-, <, > and a tolerance is used.  That would automatically cover all of those cases, and any other case really.  Without additional implementation for any custom type that supports tolerance.

#4310 I would like to discuss the possibility of resolving the #3316 fixes before closing that on master.  I'll post something over there.
